### ISSUE TITLE
Some way to prune BH

### ISSUE THREAD

**Issue by chasemc** [eddelbuettel/bh#60]
Background:
Not really an issue but I have been developing standalone "software" using shiny and a not-so-insignificant number of packages (including shiny) have BH somewhere in their dependency tree. This causes significant "bloat" to the created software and installation time.  In an impatient world, users can feel like installers are broken when they take a long time. 

Question:
First off- I am **totally against** functions/packages that modify another package. However, I am hoping there might be some way to  programmatically-prune BH to only what a calling-package actually requires. There was some talk of this here: https://github.com/stan-dev/math/issues/545 
Since I'm creating GUI-interface software that users are not developing with then I think this would be OK.

**Comment by eddelbuettel** [comment]
It is a loaded issue.

First point: R conflates 'build-time' dependency with 'run-time' dependency.  `Depends: BH` _only_ matters during compilation (and possibly if if you do later ad-hoc compilation in the `Rcpp::cppFunction()` sense, but that is rare so let me ignore it here).   If only you could declare a 'run-time' dependency you would be set.

Second point:  What is you *really* want to do?  Why would 120mb matter? I ask because sometimes with Docker and our Rocker project people go off the rail to start from a smaller Linux base (say Alpine vs Debian; or Debian vs Debian-slimmed [minus things like man pages).  That may save 100 or 200mb.  Yet all relevant container that are all-in (say, tidyverse or with texlive) are well over 2gb.  So who cares?

Third point: What is your deployment? Linux?  Are you aware of binary packages?  They could help _a lot_.

Fourth point:  Boost itself has tools like `bcp`.  We use them. But we produce BH as a _union_ of desirable packages, many added by user (ie developer request) so it becomes what it is.

Fifth point:  "Cheat".  Create an empty package BH with a higher version number. Install it once you're done building.  This approximates the run-time use (and will otherwise break your system but hey you get to keep the pieces).  

So in short: No "Free Lunch" here.  If there was a low-hanging fruit we would long have picked it.

I will stop here.  Feel free to re-focus your question.  I'll try to help but  I cannot promise that this is easily doable.

**Comment by eddelbuettel** [comment]
Can we close this if there are no further follow-ups?

**Comment by chasemc** [comment]
Thanks for the detailed reply. 
I think you're first and fifth points are really good, and I'll will definitely be looking into it. 
For me the reality isn't the size of BH, it's the time it takes to unpack the binary- I believe it's more to do with copying the large number of files and directories. I had come across `bcp` in researching this but that's over my skill level at this time. 
For now I'll just be using the binaries and be sure to include a progress window. Maybe we can have a brief discussion at CRUG sometime, as I obviously still have much to learn. 

Thanks again.
-Chase

**Comment by eddelbuettel** [comment]
What is your platform?  Is the 'unpack BH' cost one-time or repeatedly?

**Comment by chasemc** [comment]
Main concern and effort is for Windows as the majority of users will require that OS anyways. 'Unpack'/installing BH on initial install is the (small) concern, so one-time.  And especially for the "non-computer type" of users, things that take awhile "seem broken". 

Again, I don't see this as a problem with BH, I was just curious and wondered if I was missing something obvious. I am also curious if a decent majority developers are using a certain subset of the package and a parallel-derivative-subset of the package would be helpful?

I just went and researched my question and it seems like interest in something like this has grown.  I had posted the same on [RStudio community last Feb](https://community.rstudio.com/t/dplyr-dependency-on-bh-slows-installs/5294) and personally asked @hadley last March about dplyr's dependency on BH. At the time ( and this is my recollection, *not speaking for them*) their team's priority was not aligned with the work required to remove BH as a dependency. 

However, it seems that even that view has changed (or is at least public) now:       
https://github.com/tidyverse/dplyr/pull/3672
https://github.com/tidyverse/dplyr/issues/3497
https://twitter.com/hadleywickham/status/1011043021960482816

I'll note that dplyr is not my concern here.  In fact shiny itself relies on httpuv which links to BH:

https://github.com/rstudio/shiny/blob/59221dfcf235d8bdfb53435d68f947fd87a43b5e/DESCRIPTION#L68
https://github.com/rstudio/httpuv/blob/35b58a93845b7cbbb67a0079279580b04d813894/DESCRIPTION#L26

**Comment by eddelbuettel** [comment]
For Windows it is even dumber as the binaries are already built.  Just use [drat](https://github.com/eddelbuettel/drat) to provide an 'empty' BH to cope with the fact that R has no run-time depends.

I am aware of the rumblings of that team. The devil is in the detail, and splitting BH is not trivial.

**Comment by chasemc** [comment]
Awesome, thanks!
